lib/syskit/runtime/apply_requirement_modifications.rb

```diff
@@ -102,21 +102,30 @@ def syskit_apply_async_resolution_results
                     find_tasks(Syskit::InstanceRequirementsTask).running
 
                 begin
-                    return unless syskit_current_resolution.apply
+                    resolution_apply_result = syskit_current_resolution.apply
+
+                    unless resolution_apply_result.fulfilled?
+                        return resolution_apply_result
+                    end
                 ensure
                     syskit_current_resolution_keepalive.discard_transaction
                     @syskit_current_resolution = nil
                 end
 
-                running_requirement_tasks.each do |t|
+                resolution_apply_result.instances.each_key do |t|
```

Comment:
Instead of finishing InstanceRequirementsTask when the resolution finishes, my proposition would be to

- have an intermediate event `resolution_success`, which gets emitted here instead of `success`
- change the logic that finds what needs to be deployed to only pick the running InstanceRequirementsTask
- change the logic that finds whether there are new requirements to check if there are running InstanceRequirementsTask without the resolution_success event emitted (resolution_success_event.emitted?)

Reply:
Logic in runtime/apply_deployment_tasks_from_plan.rb